content/blog/2019-11-15-profile-guided-basic-block-reordering.md

```diff
@@ -0,0 +1,94 @@
+
+The algorithm first identifies frequently executed "chains", i.e., paths of basic blocks. It starts by setting all basic blocks as individual chains. Then the edges are iterated from highest weight to lowest. For each edge, if it is connecting tail and head of two chains, the chains will be merged. We implemented this by maintaining the start of chains and the next block relationships in hashmaps. Then we iterate through all the block terminators to collect edges with weights and rank them by weights. Finally we merge the chains if and only if the edge is jumping from the end of a chain to the start of a chain.
+
+After the chains are identified, they are arranged sequentially based on weights over edges cross chains. In our implementation, we decide the next chain to put in the way that the current chain has one branch with largest weight point to the start of it.
+
+
+## Evaluation
+
+### Initial Testing
+Initially we tested our pass on simple programs with only one conditional branch, where one branch is more frequently executed than the other. Our pass successfully rearranged the more used pass to directly after the block before branching. We also tested on more complicated branching scenarios.
+
+### Benchmark Testing
+
+We also evaluated the pass on [PARSEC](https://parsec.cs.princeton.edu/). Due to machine and time constraints, we only managed to run streamcluster programs with profiles provided by running the simulated inputs. We use `parsecmgmt` to run the benchmarks, and report the branch misses as well as the total runtimes. We compared clang (no optimizations) with only our optimization pass against clang (no optimizations). Unfortunately, there isn't a significant improvement in either the branch misses or the runtime. We suspect there is at least some improvement in streamcluster, considering the drastically lowered standard deviation and modestly reduced branch misses.
+
+| Benchmark                              |No optimizations     | Our pass            |
+|----------------------------------------|---------------------|---------------------|
+| Streamcluster (branch-misses 10ks)     | 3326.58094 +/- 35.9 | 3312.82899 +/- 1.45 |
+| Streamcluster (runtime)                | 15.389 +/- 1.414    | 15.316 +/- 1.704    |
+
+
+We think this is because the profiling information we collected is not enough to reflect the actual workload, since we are only using small simulated input for profiling. Clang profiling also does not provide unconditional branch frequency. In addition, the branch miss rate is already pretty low even without any optimizations. Even without any compiler optimizations, there are plenty of other optimizations in lower layers that likely allow for low branch-misses.
```

Comment:
By "small simulated input," do you mean the "simsmall" input from PARSEC? If so, that's the "simulator" input, i.e., it's small enough to make it feasible to run in an architectural simulator.

Reply:
Yes, I believe so. @q-hwang can probably weigh in more on this part.

Reply:
I believe we used `simlarge` for testing and collected profiles on all the other inputs, except for `native`, so not just `simsmall`. But all those other inputs are much smaller.